### PR TITLE
Use a recipe as an input to another recipe

### PR DESCRIPTION
**Pull request by chrisgemignani**
Expressing queries that require subselects has been difficult in recipe. This PR adds new patterns for generating and building subselects.

- Add select_from
- Adds a more general purpose find_column when building shelves from yaml
- Changes LookupDimension to support returning original values if not found in the lookup table.
- Add `isnot` filter.

## recipe.select_from()

Recipe gets a `select_from` which directly maps to a query.select_from in query generation. If a `select_from` is provided, recipe doesn't attempt to ensure only a single table is used and uses this select_from for the query.

This is a simpler approach to doing the kinds of queries that we use SummarizeOver for.

## find_column

When shelves are built from yaml, the yaml is then matched against a sqlalchemy selectable at runtime to find the columns that the sqlalchemy expressions in the ingredients should be built with. `find_column` is used to perform this lookup and it can handle selectables that are Recipes, sqlalchemy table expressions, and sqlalchemy subqueries.

## Other stuff

* Changes LookupDimension to return original values. If the `default` parameter is passed it will override this.
* An automatic filter can be built using isnot

### CODE REVIEW

File: .flake8 [hunk 1]
```diff
@@ -0,0 +1,18 @@
+[flake8]
```
Comment (radix) [inline .flake8]: We already have some flake8 configuration in setup.cfg, I'm not sure what's preferable between setup.cfg and .flake8 these days

File: recipe/shelf.py [hunk 2]
```diff
@@ -367,12 +419,17 @@ class Shelf(AttrDict):
     class Meta:
         anonymize = False
         table = None
+        select_from = None
```
[the remark this answers is not in the document]
Comment (chrisgemignani) [reply recipe/shelf.py]: It arises out of a desire for a shelf to have has dictionary-like behaviors 

```
class Shelf(AttrDict):
```

Putting attributes directly on the shelf object would treat them like ingredients.

File: recipe/core.py [hunk 3]
```diff
@@ -393,18 +396,30 @@ def query(self):
             recipe_parts = extension.modify_recipe_parts(recipe_parts)
 
         # Start building the query
-        recipe_parts['query'] = self._session.query(*recipe_parts['columns'])\
-            .group_by(*recipe_parts['group_bys']) \
-            .order_by(*recipe_parts['order_bys']) \
-            .filter(*recipe_parts['filters'])
+        if self._select_from is not None:
```
[the remark this answers is not in the document]
Comment (chrisgemignani) [reply recipe/core.py]: :+1:

File: recipe/extensions.py [hunk 4]
```diff
@@ -160,10 +160,15 @@ def add_ingredients(self):
...
-                # make a Filter and add it to filters
-                self.recipe.filters(dimension.build_filter(values, operator))
+                try:
+                    dimension = self.recipe._shelf.find(dim, Dimension)
+                    # make a Filter and add it to filters
+                    self.recipe.filters(
+                        dimension.build_filter(values, operator)
+                    )
+                except BadRecipe:
+                    # Skip if the dimension can't be found in the shelf.
```
[the remark this answers is not in the document]
Comment (chrisgemignani) [reply recipe/extensions.py]: My thinking was it was to replicate built-in Juicebox Recipe behavior, but that actually raises the same exception. 🤔

[the remark this answers is not in the document]
Comment (chrisgemignani) [reply recipe/extensions.py]: I removed this from the PR.

**Pull request merged**